[2단계 - 장바구니 기능] 후추(주찬민) 미션 제출합니다.

Combi153 opened the pull request:
안녕하세요 소니! 후추입니다 😀 

2단계 미션을 드디어 마치고 제출합니다!
일찍 제출할 수 있을 줄 알았는데, 예상하지 못했던 문제들을 많이 만났네요. 공부하고 해결하느라 시간이 걸렸습니다. 그래도 즐겁게 학습하면서 마무리할 수 있었습니다!

## 지난 1단계 코멘트

지난 1단계 리뷰에 답변을 달았습니다! 제 생각과 질문을 주로 남겼습니다. 확인해보시고 다시 코멘트 남겨주시면 감사할 것 같아요! 😁 

[1차 PR 링크](https://github.com/woowacourse/jwp-shopping-cart/pull/253)

추가로 PR 메시지에 질문하신 내용을 답변하겠습니다!

### 테스트 시 data.sql을 이용하는 것
> 추가로 JdbcTemplate 을 사용해서 데이터를 넣을 수도 있지만 data.sql을 이용해 테스트 데이터를 미리 셋팅하는 방법도 있을 것 같네요. 두 방식은 어떤 장단점이 있을까요?

JdbcTemplate을 활용하면 테스트 코드 내부에서 직접 데이터를 넣을 수 있습니다. 따라서 테스트 코드를 읽고 파악하기 용이합니다. 반면, 테스트 코드가 다소 지저분해지고 가독성이 떨어질 수 있습니다.

data.sql을 활용하면 sql 문법으로 데이터를 넣을 수 있습니다. 데이터를 넣는 과정이 편하고, 테스트 코드의 외형을 깔끔하게 유지할 수 있습니다. 반면, 테스트 내용을 이해하기 위해 data.sql 파일도 따로 확인해야 하는 번거로움이 있습니다.

개인적으로 테스트 코드를 읽을 때 data.sql 파일을 읽는 것의 비용이 더 크게 느껴집니다. 따라서 테스트코드에서 JdbcTemplate를 활용해서 데이터를 넣으려고 합니다!

한편, SimpleJdbcInsert 객체를 활용하면 JdbcTemplate의 이점을 가져가면서도 코드의 가독성을 높일 수 있습니다. 이번에 테스트를 작성하면서 SimpleJdbcInsert 를 공부해서 활용하고자 했습니다. 

### 도메인 객체의 필요성
> 지금 후추는 영속성 Entity로 사용하고 (현재는 검증밖에 없지만) 비즈니스 로직은 도메인 객체를 만들어 사용하고 있네요. 어떤 방식으로 사용할지는 개발자의 선택입니다. 이번 미션에서 후추가 1번으로 사용하면서 특별한 장점을 느끼지 못했다면 2번으로 사용해도 되지 않을까요? 그 근거는 이미 찾은것 같네요 :)

넵! 이전 단계에서 `도메인 객체`와 `DB 테이블과 매핑되는 영속성 Entity`로 구분했던 것을 `비즈니스 로직을 포함하는 도메인 Entity`로 통합하였습니다!

통합하니 코드의 가독성이 올라가고, 복잡성도 줄어든 것 같습니다. 좋은 방법을 알려주셔서 감사합니다 😃 

단, 이번에 처음 사용하는 것이다보니, 제대로 사용한 것이 맞는지 의문입니다. 소니가 보기에, 제가 Entity를 제대로 사용하고 있는지 확인해주시면 감사할 것 같습니다!!

## 집중했던 것

2단계 미션을 수행하면서 테스트를 공부하고 작성하는데 상당히 많은 시간을 들였습니다. 단위 테스트, 통합 테스트, 인수 테스트 각각이 무엇인지 알고 싶었고, 테스트 방법을 알고 싶었습니다. 그를 위해 직접 다양한 테스트를 작성해보았습니다.

이번 미션을 통해서 단위, 통합, 인수 세 가지 테스트에 대해 아래와 같이 생각을 정리했습니다.

#### 단위 테스트

- 가능한 작은 동작을 확인하는 테스트이다.
- 다른 계층 혹은 환경을 분리하여 테스트 한다.
- Mockito, @WebMvcTest 등을 활용하여 테스트하고 싶은 부분만 분리할 수 있다.
- Service, Controller, Domain 등의 테스트에서 활용할 수 있다.

#### 통합 테스트

- 전체 코드가 의도대로 동작(협력)하는지 확인하는 테스트이다.
- 단위 테스트와 달리, 실제 DB와 같이 다양한 환경과 연동해서 테스트 한다.
- Api 를 테스트하기 위해, RestAssured 등을 사용할 수 있다.

#### 인수 테스트

- 시나리오에 맞춰 수행하는 테스트이다.
- 예상 가능한 시나리오에서 코드가 의도한 동작을 하는지 확인한다.

소니가 보시기에 이러한 정리가 맞는지, 알맞게 테스트 코드를 작성했는지 확인해주시면 감사할 것 같습니다!

## 질문

### 1. 프론트엔드와 백엔드 코드가 협력하는 부분의 테스트
이번 미션에서 다음과 같은 동작을 확인하고 싶었습니다.

- 장바구니 버튼을 눌렀을 때 사용자가 선택되지 않았다면 사용자 선택화면으로 간다
- 상품 담기 버튼을 눌렀을 때 사용자가 선택되지 않았다면 사용자 선택화면으로 간다

이러한 기능은 자바스크립트로 동작합니다.
```javascript
document.addEventListener("DOMContentLoaded", function () {
        const credentials = localStorage.getItem('credentials');
        if (!credentials) {
            alert('사용자 정보가 없습니다.');
            window.location.href = '/settings';
            return;
        }
```

이러한 동작을 테스트 코드로 확인하고 싶었습니다. 인수테스트는 마땅히 이러한 동작을 확인해야 할 것이라 생각했습니다. 하지만 이를 테스트하는 방법을 모르겠습니다. 어쩌면 백엔드 개발자가 확인해야할 부분이 아닌가?라는 생각도 듭니다.

위와 같이 프론트엔드의 코드가 연관된 부분은 테스트를 해야하는지, 어떻게 할 수 있는지 궁금합니다!

### 2. 예외 처리
이번에 @ControllerAdvice, @ExceptionHandler를 통해 예외 처리하는 방법을 공부했습니다. 개발자가 예상가능한 예외를 미리 처리해두면, 사용자에게 적절한 예외 응답을 할 수 있다고 생각했습니다. 그래서 예외를 다룰 필요성을 느꼈습니다.

하지만, **사용자가 요청할 수 없는 것 같지만, url로 요청이 올 수 있는 예외인 경우**는 예외 처리를 해야할 지 궁금합니다. 가령, 존재하지 않는 productId 값을 삭제하라는 요청이 온 경우입니다.

사용자는 웹 애플리케이션 내부에서 productId에 접근할 방법이 없으므로, 해당 예외가 발생할 가능성이 없습니다. 다만 다음과 같이 url로 동작을 요청할 수는 있습니다.
![image](https://user-images.githubusercontent.com/106813090/236217199-4413fd6e-d1c2-4325-b928-9cc9541efab9.png)

**사용자가 요청할 수 없는 것 같지만, url로 요청이 올 수 있는 예외인 경우** 처리를 해야할까요? 소니는 예외를 어디까지 예상하고 처리하시나요? 소니의 생각이 궁금합니다!!!

리뷰 잘 부탁드립니다! 감사합니다🙂🙂

Combi153 committed:
refactor: dto에 기본 생성자 추가

Combi153 committed:
refactor: Service 메서드 반환값 수정 및 단위 테스트 추가

Combi153 committed:
style: import 순서 변경

Combi153 committed:
test: 테스트 코드에 Dao 의존성 제거 및 AfterEach 추가

Combi153 committed:
refactor: Controller 관심사 분리 및 이름 수정

Combi153 committed:
docs: 2단계 기능 요구 사항 추가

Combi153 committed:
feat: 비밀번호 기능 추가

Combi153 committed:
docs: email에 대한 기능 요구사항 추가

Combi153 committed:
feat: Email 기능 추가

Combi153 committed:
refactor: 패키지 구조 변경

Combi153 committed:
feat: User 기능 추가

Combi153 committed:
docs: 기능 요구사항 추가

Combi153 committed:
feat: DDL 수정 및 추가

Combi153 committed:
docs: DB 특이사항 기록

Combi153 committed:
refactor: DDL 수정

변경 내용:
- user 테이블 이름을 users로 수정
- h2 데이터베이스 예약어로 인한 이슈

Combi153 committed:
refactor: ProductDao 코드 수정

변경 내용 :
- 메서드 이름 변경
- RowMapper 객체 상수화
- Column 내용 상수화

Combi153 committed:
feat: UserEntity 추가

Combi153 committed:
feat: UserDao 기능 추가

Combi153 committed:
refactor: CartService 이름을 ProductService 로 변경

Combi153 committed:
refactor: dto <-> domain, dto <-> entity 변환 로직 수정

변경 내용:
- dto <-> domain : dto에서 domain을 생성할 수 있도록 수정
- dto <-> entity : dto에서 entity의 포장을 해제할 수 있도록 수정

Combi153 committed:
feat: UserService 기능 추가

Combi153 committed:
refactor: Controller 이름 변경

변경 내용:
- CartApiController -> ProductApiController 로 이름 변경
- CartController -> ViewController 로 이름 변경

Combi153 committed:
refactor: 테스트 로직 수정

Combi153 committed:
feat: 사용자 설정 페이지 연동

Combi153 committed:
refactor: @Autowired 삭제

Combi153 committed:
style: 파일 끝 개행 추가

Combi153 committed:
docs: 기능 요구사항 추가 및 정리

Combi153 committed:
feat: CartDao 기능 추가

Combi153 committed:
feat: CartService 기능 추가

Combi153 committed:
feat: UserDao 이메일로 데이터 조회 기능 추가

Combi153 committed:
feat: UserService 이메일로 id 조회 기능 추가

Combi153 committed:
feat: ProductService id로 상품 이름, 가격, 이미지 주소 조회 기능 추가

Combi153 committed:
feat: 장바구니 화면 연동

Combi153 committed:
feat: 장바구니 내역 읽기 기능 추가

Combi153 committed:
feat: CartService 장바구니에 상품을 넣는 기능 추가

Combi153 committed:
feat: CartApiController 장바구니에 상품을 넣는 기능 추가

Combi153 committed:
feat: CartDao에 데이터 삭제 기능 추가

Combi153 committed:
feat: CartService에 장바구니 상품 삭제 기능 추가

Combi153 committed:
feat: CartApiController 장바구니 상품 제거 기능 추가

Combi153 committed:
refactor: 도메인 용어 User를 Member 로 수정한다

변경 이유 :
- h2 데이터베이스 예약어 user 로 인한 변경

Combi153 committed:
docs: 테스트 관련 기록 추가

Combi153 committed:
test: ProductApiController 단위 테스트 추가

Combi153 committed:
test: CartApiController 단위 테스트 추가

Combi153 committed:
test: ViewController 단위 테스트 추가

Combi153 committed:
refactor: 상품 수정 기능 리팩터링

변경 내용:
- ProductDao 수정 메서드 파라미터 순서 변경
- ProductService 수정 메서드 파라미터 추가
- ProductApiController PathVariable id 값 추가

Combi153 committed:
test: ProductApiController 통합 테스트 코드 리팩터링 및 추가

Combi153 committed:
test: ProductApiController 단위 테스트 수정

Combi153 committed:
test: CartApiController 단위 테스트 수정

Combi153 committed:
feat: argumentResolver 적용

Combi153 committed:
feat: 인증 예외 처리 추가

Combi153 committed:
test: CartApiController 통합 테스트 추가

Combi153 committed:
test: 인수 테스트 추가

Combi153 committed:
refactor: 패키지 구조 변경

Combi153 committed:
docs: 구현한 기능 체크 및 요구사항 추가

Combi153 committed:
feat: 예외 처리 기능 추가

Combi153 committed:
refactor: 도메인 객체와 Entity를 하나로 리팩터링

Combi153 committed:
refactor: 테스트 전용 sql 설정

Combi153 committed:
refactor: 예외 처리 로직 변경

Combi153 committed:
refactor: @Transactional 적용

sonypark commented on src/main/java/cart/controller/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+
+public class AuthenticationPrincipalArgumentResolver implements HandlerMethodArgumentResolver {
+
+    @Override
+    public boolean supportsParameter(final MethodParameter parameter) {
+        return parameter.hasParameterAnnotation(AuthenticationPrincipal.class);
+    }
+
+    @Override
+    public Object resolveArgument(final MethodParameter parameter, final ModelAndViewContainer mavContainer, final NativeWebRequest webRequest, final WebDataBinderFactory binderFactory) {
+        final String authorization = webRequest.getHeader("authorization");
+
+        if (authorization == null || !authorization.toLowerCase().startsWith("basic")) {
+            throw new IllegalArgumentException();
```
- ArgumentResolver의 역할은 무엇일까요?
- 현재는 ArgumentResolver에서 인증에 대한 처리도 해주고 있는데 인증까지 ArgumentResolver가 수행할 책임에 들어갈까요?
- 인증로직을 구현할때 Spring Interceptor도 고려하셨는지도 궁금합니다 :)

Combi153 replied:
- ArgumentResolver 는 Controller 메서드의 매개변수에 데이터를 바인딩하는 역할을 할 수 있습니다.
- 말씀하신대로, 현재 ArgumentResolver는 데이터를 바인딩할 뿐 아니라 인증에 대한 처리까지 해주고 있네요. 과도한 책임을 지운 것 같습니다!
- Interceptor 을 자세히 몰라서 고려하지 못했습니다. 이번에 공부해서 적용해보겠습니다!

Interceptor는 Controller 호출 전 후에 요청과 응답을 가로채서 공통 작업을 수행하는 객체입니다. 
ArgumentResolver 를 통해 Controller 메서드의 매개변수에 데이터를 바인딩하고, Interceptor를 활용해서 인증을 확인하겠습니다!

sonypark commented on src/main/java/cart/controller/Credentials.java:
```diff
@@ -0,0 +1,20 @@
+package cart.controller;
+
+public class Credentials {
+
+    private final String email;
+    private final String password;
```
현재는 email,password가 null이거나 의도치 않은 값이 들어올 수 있겠네요.
- 검증로직이 필요하지 않을까요?
- 필요하다면 값객체도 고려해보는건 어떨까요?

Combi153 replied:
검증로직이 필요하다고 생각합니다! 짚어주셔서 감사합니다!

-  null 혹은 빈 값인지 확인해봐야 할 것 같습니다. 
- 다만 값 객체가 필요할만큼 큰 기능은 아니라고 생각했습니다. 값 객체를 두는 곳이 오히려 복잡성을 증가시킬 것 같았습니다. 따라서 AuthInfo의 생성자에서 검증을 수행하도록 했습니다.

혹시 소니가 값객체를 권유해주신 특별한 이유가 있으셨을까요?

sonypark commented on src/main/java/cart/controller/AuthenticationPrincipal.java:
```diff
@@ -0,0 +1,13 @@
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface AuthenticationPrincipal {
+
+    boolean required() default true;
+}
```
인증 관련된 클래스는 authentication 과 같은 패키지로 구분해도 좋을 것 같아요

Combi153 replied:
넵 패키지를 적절히 분리해보겠습니다! 감사합니다 😀

sonypark commented on src/main/java/cart/domain/MemberEntity.java:
```diff
@@ -0,0 +1,30 @@
+package cart.domain;
+
+public class MemberEntity {
+
+    private final Long id;
+    private final Email email;
+    private final Password password;
+
+    public MemberEntity(final Long id, final String address, final String password) {
+        this.id = id;
+        this.email = new Email(address);
+        this.password = new Password(password);
+    }
+
```
값객체 사용 💯
애초에 MemberEntity에서 Email, Password를 받는건 어떤가요?

```suggestion

    private final Long id;
    private final Email email;
    private final Password password;

    public MemberEntity(final Long id, final Email email, final Password password) {
        this.id = id;
        this.email = email;
        this.password = password;
    }

```

Combi153 replied:
감사합니다! Email과 Password를 객체로 적절히 만든 것 같아 뿌듯하네요!

제가 MemberEntity에게 Email과 Password 생성 책임을 둔 이유는 다음과 같습니다.

- 회원이 생성될 때, Email과 Password가 함께 생성되는 것이 맞다고 생각했습니다. 회원이 존재하는 경우에만, 그 회원의 Email과 Password가 존재하는 것이 합당하다고 느껴졌기 때문입니다. 만약 MemberEntity 밖에서 Email과 Password가 생성된다면, 회원은 존재하지 않는데 Email과 Password가 생성된 경우가 있게 됩니다. 이것이 각 객체의 맥락과 맞지 않게 느껴졌습니다.
- GRASP 의 Creator 원칙이 이 상황에 맞다고 생각했습니다. Email과 Password는 MemberEntity 객체의 일부입니다. 따라서 MemberEntity 객체에게 생성 책임을 부여했습니다.

아직 미숙해서 시야가 좁습니다. 혹시 MemberEntity에서 Email, Password를 받는 것을 권유해주신 특별한 이유가 있으실까요? 알려주신다면 공부하고 고민해보겠습니다!! 😀

sonypark commented on src/main/java/cart/domain/Password.java:
```diff
@@ -0,0 +1,45 @@
+        validateFormat(value);
+        this.value = value;
+    }
+
+    private void validateFormat(final String value) {
+        if (value.length() < MINIMUM_LENGTH || value.length() > MAXIMUM_LENGTH) {
+            throw new IllegalArgumentException("비밀번호는 최소 " + MINIMUM_LENGTH + "자 이상, 최대 " + MAXIMUM_LENGTH + "자 이하여야 합니다.");
+        }
+        if (notContains(NUMBER_PATTERN, value)) {
+            throw new IllegalArgumentException("비밀번호는 숫자를 포함해야 합니다.");
+        }
+        if (notContains(LETTER_PATTERN, value)) {
+            throw new IllegalArgumentException("비밀번호는 대소문자를 포함해야 합니다.");
+        }
+        if (notContains(SPECIAL_LETTER_PATTERN, value)) {
+            throw new IllegalArgumentException("비밀번호는 특수문자를 포함해야 합니다.");
+        }
+    }
```
꼼꼼한검증💯

Combi153 replied:
만들고 나니 뿌듯했어요 감사합니다 😀

sonypark commented on src/main/java/cart/domain/Password.java:
```diff
@@ -0,0 +1,45 @@
+    private static final Pattern SPECIAL_LETTER_PATTERN = Pattern.compile("[~`!@#$%^&*()-+=]");
+
+    private final String value;
+
+    public Password(final String value) {
+        validateFormat(value);
+        this.value = value;
+    }
```
- 사용자가 입력한 password를 그대로 DB에 저장하면 어떤 이슈가 있을까요?
- 이슈가 있다면 어떻게 해결할 수 있을까요?

Combi153 replied:
미션 진행하는 동안은 한번도 생각해보지 못했던 부분인데, 키워드를 알려주셔서 감사합니다. 재미있는 주제였어요!

- 보안이 취약한 것이 문제일 것 같습니다! DB가 외부에 드러나는 순간, 비밀번호가 그대로 유출될 수 있습니다.
- 단방향 해시함수를 통해서 password를 암호화할 수 있을 것입니다. 임의의 데이터를 추가하거나(솔팅), 해시함수를 반복해서 적용함으로써 보안성을 향상시킬 수도 있습니다.
- 다만 지금은 회원가입 기능 등을 직접 만든 것이 아니기에 바로 적용하기 어려운 것 같습니다. 다음에 회원가입 기능을 만들게 되면 그때는 꼭 적용해보겠습니다. 😊

평소에 네이버나 구글 등에서 비밀번호 찾기를 하면, 왜 비밀번호를 알려주지 않고 재발급 받게 하는지 그 이유를 알 것 같습니다! 😀

sonypark commented on src/main/java/cart/domain/Password.java:
```diff
@@ -0,0 +1,45 @@
+
+public class Password {
+
+    private static final int MINIMUM_LENGTH = 8;
+    private static final int MAXIMUM_LENGTH = 20;
+    private static final Pattern NUMBER_PATTERN = Pattern.compile("\\d+");
+    private static final Pattern LETTER_PATTERN = Pattern.compile("[a-zA-Z]");
+    private static final Pattern SPECIAL_LETTER_PATTERN = Pattern.compile("[~`!@#$%^&*()-+=]");
```
`Pattern.compile()` 사용 👍

Combi153 replied:
Pattern 객체가 비싸다는 얘기를 들어서 적용해보았습니다. 감사합니다! 😁

sonypark commented on src/main/java/cart/domain/CartEntity.java:
```diff
@@ -0,0 +1,30 @@
+package cart.domain;
+
+public class CartEntity {
+
+    private final Long id;
+    private final Long memberId;
+    private final Long productId;
+
+    public CartEntity(final Long id, final Long memberId, final Long productId) {
+        this.id = id;
+        this.memberId = memberId;
+        this.productId = productId;
+    }
+
```
Member, Product의 객체가 아닌 id를 정의하셨는데 이유가 있을까요~?
설계 과정에서 DB 설계를 먼저 하셨는지. 도메인 설계를 먼저하셨는지도 궁금합니다 :)

Combi153 replied:
DB를 먼저 설계했습니다. 그 이유는 도메인에 설계할 만한 비즈니스 로직이 없다고 생각이 들었기 때문입니다. 그래서 DB의 구조가 도메인에 그대로 들어간 것 같습니다. 하지만 레벨 1동안 연습한 객체지향을 버리면 안 되죠. 반성하겠습니다!

CartEntity가 Member, Product 객체를 상태로 가질 수 있도록 리팩터링 해보겠습니다. 그러면서 이전의 코드와 바뀐 코드의 장단점을 비교해보겠습니다!

sonypark commented on src/test/java/cart/controller/unit/CartApiControllerUnitTest.java:
```diff
@@ -0,0 +1,75 @@
+        //expect
+        mockMvc.perform(get("/carts")
+                        .header("Authorization", "Basic " + ENCODED_CREDENTIALS)
+                        .accept(MediaType.APPLICATION_JSON))
+                .andExpect(jsonPath("$[0].id").value(equalTo(1)))
+                .andExpect(jsonPath("$[0].name").value(equalTo("치킨")))
+                .andExpect(jsonPath("$[0].price").value(equalTo(10000)))
+                .andExpect(jsonPath("$[0].image").value(equalTo("치킨 주소")));
+    }
```
jsonPath 대신 ProductEntity 객체로 파싱한뒤 값을 검증하면 가독성이 조금 더 좋아질 것 같네요 :)

Combi153 replied:
넵 객체로 파싱해서 가독성을 높여보겠습니다!

ProductEntity 객체로 파싱하길 권하셨지만, 두 가지 이유로 다른 선택을 내렸습니다.

- 특정 객체로 파싱하기 위해 ObjectMapper 를 사용하는데 그를 위해서는 기본생성자가 필요합니다. 기본생성자를 정의하면, 필드에 final 키워드를 붙일 수 없게 됩니다. ProductEntity 는 도메인 객체이므로 필드에 final 키워드를 붙여 불변 객체로 사용하고 싶습니다. 따라서 ProductEntity 보다는 다른 객체로 파싱하는 것이 좋다고 판단했습니다.
- 현재 테스트하고 있는 메서드 readCart()는 List\<ResponseProductDto>를 반환합니다. 현재는 ResponseProductDto와 ProductEntity의 필드가 동일하지만, 앞으로도 그럴 것이라 보장하기 힘듭니다. 따라서 ProductEntity 보다는 ResponseProductDto로 파싱하는 것이 적절할 것이라 생각했습니다. 한편, ResponseProductDto로 파싱하여도 가독성을 높이는데는 충분한 역할을 할 것이라 생각했습니다.

결론적으로, ResponseProductDto로 파싱해서 값을 검증하도록 코드를 수정했습니다. 😃 

코드를 수정하면서, json 값을 꺼내어 역직렬화 할 때, 한글이 있는 경우 인코딩 문제가 발생하는 것을 확인했습니다. 가령, `"치킨"`이 자꾸만 `"ì¹í¨"`으로 나왔습니다. 구글링해서 나름 설정을 바꾸어봤지만, 해당 이슈를 해결할 수 없어 테스트 케이스를 영어로 작성했습니다.

sonypark commented on src/main/java/cart/controller/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public Object resolveArgument(final MethodParameter parameter, final ModelAndViewContainer mavContainer, final NativeWebRequest webRequest, final WebDataBinderFactory binderFactory) {
+        final String authorization = webRequest.getHeader("authorization");
+
```
Spring에서 제공하는 `HttpHeaders.AUTHORIZATION` 을 사용해보면 좋을 것 같아요~

Combi153 replied:
미리 선언된 상수가 있었네요! 알려주셔서 감사합니다 😄

sonypark commented on src/test/java/cart/controller/unit/ProductApiControllerUnitTest.java:
```diff
@@ -0,0 +1,76 @@
+
+    private Map<String, String> valueByFields;
+
+    @BeforeEach
+    void setUp() {
+        valueByFields = new HashMap<>();
+    }
+
+    @Test
+    void 상품을_생성한다() throws Exception {
+        // given
+        valueByFields.put("name", "치킨");
+        valueByFields.put("price", "10000");
+        valueByFields.put("image", "치킨 주소");
+
```
HashMap에 값을 담기보다 정의된 DTO 객체에 담는게 좋지 않을까요? (RequestCreateProductDto, RequestUpdateProductDto)

Combi153 replied:
그런 방법이 있었군요!! 훨씬 깔끔해지는 것 같습니다 감사해요 😄

sonypark commented on src/test/java/cart/controller/integration/CartApiControllerTest.java:
```diff
@@ -0,0 +1,176 @@
+        cartJdbcInsert = new SimpleJdbcInsert(dataSource).withTableName("cart");
+        jdbcTemplate = new JdbcTemplate(dataSource);
+    }
+
+    @Nested
+    class ReadTest {
+
```
Nested 테스트 👍

sonypark commented on src/test/java/cart/controller/integration/CartApiControllerTest.java:
```diff
@@ -0,0 +1,176 @@
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.Matchers.*;
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Sql("classpath:data-test.sql")
+class CartApiControllerTest {
+
+    private static final String ENCODED_CREDENTIALS = "aHVjaHVAd29vd2FoYW4uY29tOjEyMzQ1NjdhIQ=="; // huchu@woowahan.com:1234567a!
+    private static final String UNAUTHORIZED_ENCODED_CREDENTIALS = "bm9Ad29vd2FoYW4uY29tOjEyMzQ1NjdhIQ=="; // no@woowahan.com:1234567a!
+    private static final SqlParameterSource PRODUCT_PARAMS = new MapSqlParameterSource()
+            .addValue("name", "치킨")
+            .addValue("price", 1_000)
+            .addValue("image", "치킨 사진");
+    private static final SqlParameterSource MEMBER_PARAMS = new MapSqlParameterSource()
+            .addValue("email", "huchu@woowahan.com")
+            .addValue("password", "1234567a!");
+
+    @LocalServerPort
+    int port;
+
+    @Autowired
+    DataSource dataSource;
+
+    private SimpleJdbcInsert productJdbcInsert;
+    private SimpleJdbcInsert memberJdbcInsert;
+    private SimpleJdbcInsert cartJdbcInsert;
+    private JdbcTemplate jdbcTemplate;
```
통합테스트 💯

sonypark commented on src/main/java/cart/service/MemberService.java:
```diff
@@ -0,0 +1,37 @@
+        try {
+            final MemberEntity memberEntity = memberDao.findByEmail(email);
+            return memberEntity.getId();
+        } catch (EmptyResultDataAccessException e) {
+            throw new AuthorizationException(email + " 계정의 회원은 존재하지 않습니다.");
+        }
```
AuthorizationException는 email만 인자로 받고 상세 error message는 내부에 있어야할것 같네요

Combi153 replied:
훨씬 좋은 것 같습니다! 덕분에 코드가 깔끔해집니다!!

sonypark commented on src/test/java/cart/controller/unit/ViewControllerUnitTest.java:
```diff
@@ -0,0 +1,104 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+@WebMvcTest(ViewController.class)
+class ViewControllerUnitTest {
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private ProductService productService;
+
+    @MockBean
+    private MemberService memberService;
+
```
html 부분까지 테스트를 하셨네요. (디테일.. 👍)
Veiw 관련 부분은 테스트 작성 우선순위에서는 후순위이기 때문에 작성 안해주셔도 될 것 같습니다.(현재는 서버에서 view rendering까지 하고 있지만 이런 경우는 실무에서 거의 없을 거기 때문에..)
더 중요한 서버의 비즈니스 로직이 담긴 단위테스트, 통합테스트에만 꼼꼼히 작성해도 충분할 것 같아요!
물론 작성하지 말라는건 아닙니다. 후추의 에너지를 아껴 더 중요한 테스트와 비즈니스 로직에 집중해보자는 의견을 남겨봅니다!😁

Combi153 replied:
넵 조언 감사드립니다! 이번에는 테스트를 공부하고 싶어서 무리해서 여러 테스트를 작성한 것 같습니다. 말씀해주신 것처럼 테스트를 작성하는데 드는 에너지와 시간 모두 비용이므로, 앞으로는 우선순위에 초점을 맞춰 효율적으로 개발하겠습니다!! 😄

sonypark reviewed (changes requested):
안녕하세요 후추, 리뷰어 소니 입니다.😊
2단계 미션도 멋지게 구현해주셨네요👍
특히 통합 테스트, 단위 테스트를 꼼꼼히 작성해주셔서 너무 좋았습니다💯
질문주신 부분과 코드상에 피드백 드릴 부분에 대해 코멘트 남겼으니 확인 부탁드립니다 🙏 

> 프론트엔드와 백엔드 코드가 협력하는 부분의 테스트

네 후추가 말씀하신부분에 저도 동의합니다. 이 부분은 서버에서 검증은 힘들것 같아요. 서버에서는 인증에 대한 검증만 하고 화면의 리다이렉션 관련한 테스트는 한다면 프론트 쪽에서 해야할 것 같네요.

> 예외 처리

네 우선 예측가능한 예외에 대해서는 모두 처리를 해야합니다. 그리고 나머지 예측 불가능하거나 대응하기 어려운 예외는 최상위 Exception으로 받아서 서버 에러를 던져야할 것 같습니다. 예시로 드신 부분의 경우는 예측가능한 에러이고 존재하지 않는 ID에 대한 요청이기 때문에 400 Bad Request에러로 보고 처리할 수 있을 것 같네요 :)

Combi153 committed:
refactor: 클래스 이름 변경 및 패키지 구조 수정

변경 내용 :
- Credentials 클래스 이름 AuthInfo 로 변경
- 인증 관련 패키지 생성 및 구분

Combi153 committed:
feat: 도메인 객체 동등성 부여

Combi153 committed:
refactor: CartEntity 객체 상태 변경

변경 내용 :
- CartEntity 객체가 MemberEntity, ProductEntity 를 상태로 갖도록 수정
- 도메인 객체 상태 변화에 따른 Dao, Service, Controller 코드 변경

Combi153 committed:
style: final 키워드 추가

Combi153 committed:
feat: Basic credentials 디코딩 객체 BasicAuthorizationExtractor 추가

Combi153 committed:
feat: Interceptor 기능 추가 및 인증 검증 로직 수정

Combi153 committed:
feat: 인증 검증 로직 추가

Combi153 committed:
refactor: 상품 수정 및 삭제 검증 로직 Service에서 Dao로 변경

Combi153 committed:
refactor: 메서드 삭제 및 테스트 코드 추가

Combi153 committed:
refactor: MemberService 테스트 코드 추가 및 메서드 Transactional 애노테이션 추가

Combi153 committed:
refactor: 상품 관련 uri 변경

Combi153 committed:
test: HashMap을 DTO로 수정

Combi153 committed:
test: 테스트 코드 검증 로직 변경

Combi153 committed:
test: 테스트 코드 메서드 이름 수정

Combi153 committed:
feat: 예외 처리 로직 추가 및 Logger 적용

Combi153 committed:
feat: AuthInfo 예외 처리 로직 추가

Combi153 committed:
feat: 인증 예외 처리 로직 추가 및 테스트 코드 작성

Combi153 committed:
feat: 예외 처리 로직 추가

Combi153 commented:
안녕하세요 소니! 후추입니다 😀

한번 더 2단계 미션 제출합니다!
소니의 리뷰를 받고 또 여러가지 시도를 해보고 공부를 할 수 있었습니다.
좋은 리뷰에 정말 감사드립니다!! 🙇‍♂️ 

## 도전과 리팩터링
리뷰를 받은 후 코드를 고치면서 아래와 같은 사항을 공부하고, 코드에 적용했습니다.

- logger 추가
- 예외 처리 로직 추가
- 인증 관련 테스트 코드 추가
- 인증 관련 검증 로직 추가
- interceptor로 인증 로직 수행
- uri 수정 (RESTful) 

## 질문
### 1. DefaultHandlerExceptionResolver와 예외 처리
이번에 @ControllerAdvice와 @ExceptionHandler를 통해 여러가지 예외를 처리해보았습니다. 그중 "HttpRequestMethodNotSupportedException", "HttpMessageNotReadableException" 등과 같은 예외가 있었습니다. 

그런데 이러한 예외 상황을 하나씩 살펴볼 때, DefaultHandlerExceptionResolver가 예외 처리를 해주는 것을 보았습니다. 아래와 같이 body에 적절한 정보를 담았고, 상태코드는 Bad Request를 반환해주었습니다. 또한 로그도 적절히 찍어주었습니다.

```
{
    "timestamp": "2023-05-08T06:33:10.363+00:00",
    "status": 400,
    "error": "Bad Request",
    "path": "/products/1"
}
```

그렇다면, DefaultHandlerExceptionResolver 가 처리해주는 예외를 개발자가 직접 처리할 필요성이 있을지 궁금증이 생깁니다. 특히나 HttpMessageNotReadableException 등과 같은 예외는 사용자가 마주하기보다는 개발자들이 마주할 일이 많은 예외라고 생각이 들기 때문입니다.

소니는 이러한 예외를 어떻게 처리하시는 편이신지 궁금합니다!

### 2. 코드의 중복 제거
인증 로직이 있는 CartApiController를 개발하면서 ArgumentResolver, Interceptor 가 중복 코드를 줄여주었습니다. 하지만 현재도 중복 코드가 남아있습니다.

![image](https://user-images.githubusercontent.com/106813090/236753977-f7bb2e6e-0d57-4fae-a2b9-f6b1d59ba077.png)
(클릭하시면 잘 보입니다!!)

```java
final MemberEntity member = memberService.findByEmail(authInfo.getEmail());
```
이러한 코드의 중복도 ArgumentResolver, Interceptor를 활용하면 줄여볼 수 있지 않을까 생각이 듭니다. 다만 ArgumentResolver, Interceptor에게 너무 많은 책임을 부여하는 것 같다는 생각도 듭니다.

소니라면 이러한 중복 코드를 어떻게 해결하실 것인지 궁금합니다!

### 3. Controller와 Service
현재 CartApiController는 다양한 Service를 의존하고 활용합니다. 가령, 장바구니에 상품을 추가하는 createCart 메서드는 MemberService로 회원을 조회하고, ProductService로 상품을 조회하고, CartService로 장바구니에 데이터를 넣습니다.

Controller의 한 가지 메서드를 처리하기 위해 여러 개의 Service를 호출하게 되면, 애플리케이션이 커질 수록 코드를 읽기 어려워지고, 한 메서드가 너무 다양한 일을 하게 될 것 같습니다.

대안으로는 CartService에서 MemberService, ProductService 등을 의존하는 방법을 생각했습니다. 다만 의존의 방향이 적절하지 않다는 생각이 들었습니다. Service란 특정 도메인에 관한 비즈니스 로직을 담게 되는데, 이러한 로직이 의존하는 다른 Service의 변화에 영향을 받게 될 가능성이 있기 때문입니다.

결국 스스로 답을 내리지 못했습니다. 현재 CartApiController가 다양한 Service를 의존하는 것을 어떻게 리팩터링 해볼 수 있을까요?

지난 리뷰 정말 감사했습니다. 이번에도 잘 부탁드립니다! 😀 😀 😀

sonypark commented on src/main/java/cart/controller/authentication/AuthInfo.java:
```diff
@@ -0,0 +1,36 @@
+            throw new AuthorizationException("회원 비밀번호가 입력되지 않았습니다.");
+        }
+    }
+
```
request에 있는 email, password 값을 검증할때 만들어두신 Email, Password 값객체를 쓰는건 어떨까요?
이렇게 말씀드린 이유는 아래와 같습니다.
- 불완전한 검증: 현재 AuthInfo에서 하는 검증은 값이 존재하는지 여부만 판단할뿐 이메일 형식이나 비밀번호 형식에 관한 검증은 하지 않습니다. 이는 불완전한 검증으로 결국 뒤에서 또다시 검증을 해야함을 의미합니다.
- 예외는 가능한 앞단에서 빨리 발생시키는것이 좋다. 앞단에서 발생시킬 수 있는 예외를 도메인 레이어까지 끌고 갈 필요는 없을 것 같습니다. 성능면이나 코드 품질면에서도 사용자가 잘못 입력한 값에 대한 예외는 빠르게 throw해주는게 좋다고 생각합니다.

여기서 Email, Password라는 값객체로 매핑한다면 이후 사용될 Email, Password는 검증이 완료된 값이기 때문에 추가 검증없이 사용할 수 있어 사용성이 높아질 것 같습니다. (MemberEntity에서 생성인자로 값객체를 받는 부분도 동일한 이유로 말씀드렸습니다 😃)

Combi153 replied:
불완전한 검증과 빠른 예외 발생에 대한 키워드를 주셔서 감사합니다.

오래 고민했고 결국 저만의 답을 내렸습니다. 그 과정은 아래와 같습니다.

## 처음 생각
AuthInfo 에서 Email, Password 값 객체를 쓰고 싶지 않았습니다. 그 이유는 회원가입 시 예외와 인증 시 예외를 구분하고 싶었기 때문입니다. 

네이버 포털에서 회원가입과 로그인으로 예시를 들어보겠습니다.

<img width="561" alt="image" src="https://github.com/woowacourse/jwp-shopping-cart/assets/106813090/423bd1d0-a7e6-4591-8dd8-77da9f14f9c0">

네이버에서 회원가입을 할 때 비밀번호 조건을 만족하지 않는다면 `8~16자 영문 대 소문자, 숫자, 특수문자를 사용하세요.` 라는 경고 메시지를 띄웁니다. 

이것이 회원가입 시 검증하고, 응답해야 하는 예외 메시지라고 생각했습니다.

<img width="546" alt="image" src="https://github.com/woowacourse/jwp-shopping-cart/assets/106813090/0aeccd7a-d0bc-46c9-827e-ebe3605a8f38">

반면 네이버에서 로그인을 하면서 잘못된 비밀번호를 입력할 경우 `아이디(로그인 전용 아이디) 또는 비밀번호를 잘못 입력했습니다. 입력하신 내용을 다시 확인해주세요.`라는 경고 메시지를 띄웁니다.

이것이 인증 시 검증하고, 응답해야 하는 예외 메시지라고 생각했습니다.

즉, 회원가입과 인증 시 각각 처리해야 하는 예외와 응답이 다르다고 생각했습니다.

하지만 AuthInfo에서 Password 를 값 객체로 사용할 경우 인증 시 의도하지 않은 예외를 발생시키게 됩니다. Password에서 회원가입 비밀번호 조건을 검증하고 예외를 발생시키고 있기 때문입니다.

```java
public class Password {

    // ...

    private final String value;

    public Password(final String value) {
        validateFormat(value);
        this.value = value;
    }

    private void validateFormat(final String value) {
        if (value.length() < MINIMUM_LENGTH || value.length() > MAXIMUM_LENGTH) {
            throw new IllegalArgumentException("비밀번호는 최소 " + MINIMUM_LENGTH + "자 이상, 최대 " + MAXIMUM_LENGTH + "자 이하여야 합니다.");
        }
        if (notContains(NUMBER_PATTERN, value)) {
            throw new IllegalArgumentException("비밀번호는 숫자를 포함해야 합니다.");
        }
        if (notContains(LETTER_PATTERN, value)) {
            throw new IllegalArgumentException("비밀번호는 대소문자를 포함해야 합니다.");
        }
        if (notContains(SPECIAL_LETTER_PATTERN, value)) {
            throw new IllegalArgumentException("비밀번호는 특수문자를 포함해야 합니다.");
        }
    }

    private boolean notContains(final Pattern pattern, final String value) {
        final Matcher matcher = pattern.matcher(value);
        return !matcher.find();
    }

    // ...
```

## 생각 전환
한편, 소니께서 말씀해주신 불완전한 검증으로 인한 문제가 무엇인지 고민했습니다.

애초에 header에 담긴 인증 정보가 Password 객체의 검증 로직을 통과하지 못한다면, 회원일리 만무했습니다. 따라서 미리 Password 객체로 인증 정보를 검증한다면, DB에서 회원 여부를 굳이 확인하지 않고 빠르게 예외를 던질 수 있었습니다. 이것이 값 객체를 사용하는 것의 매우 큰 이점이었습니다.

## 결론
### AuthInfo 에서 Email, Password 값 객체를 사용하도록 리팩터링했습니다.

다만, 회원가입과 인증 시 예외를 구분하기 위해 아래와 같은 메서드를 추가했습니다.

```java
private static AuthInfo generateAuthInfo(final String email, final String password) {
        try {
            return new AuthInfo(email, password);
        } catch (IllegalArgumentException exception) {
            throw new AuthorizationException("입력한 정보의 회원은 존재하지 않습니다.");
        }
    }
```

이 메서드를 추가함으로써 인증 시 잘못된 인증 정보에 대해 AuthorizationException이 발생하게 되고, 적절히 응답할 수 있게 됩니다.

소니께서 값객체를 사용해보라고 권유해주신 이유를 제가 적절히 받아들였는지 궁금합니다 😀

sonypark replied:
네 인증로직을 기계적으로 구현하지 않고, 이렇게 깊게 생각해보고 정리하는게 참 좋은 것 같습니다.👍

후추 말대로 회원가입을 할때의 검증과, 로그인할 때의 검증은 다릅니다. 그리고 분리되어야 하는게 맞다고 생각합니다. 예외 메시지도 달라야하는것에 동의합니다 :) 👍👍
인터셉터에서 값객체를 사용하게 되면 말씀하신대로 회원가입시 발생할 패스워드 검증에 대한 예외메시지가 발생하겠네요. 그 이유라면 값객체를 사용하지 않고 not null체크 not empty 정도의 체크만 하거나 대안으로 말씀하신대로 값객체 사용후 try-catch로 예외를 만들어 던져주는 방법이 있을것 같습니다. 저라면 이렇게 예상 가능한 예외는 앞단에서 최대한 빨리 던져주는것을 선호해서 후자로 할 것 같긴 합니다😃

FYI: 서비스에서 회원가입시 인증에 대한 빈도와 로그인 시 또는 로그인 후 사용자에 대한 인증에 대한 빈도를 비교해보자면, 전자는 최초 1회에만 해당하고 그 이후에는 모두 후자에 대한 인증이 필요할 겁니다. 따라서 인터셉터에서는 회원가입을 제외한 사용자 인증이 필요한 부분에 대해서 검증 로직을 넣어주고, 회원가입에 대한 요청은 도메인에서만 예외처리를 할 수 있을 것 같습니다.

아래와 같이 `excludePathPatterns` 키워드를 사용하면 인터에서 해당 path로 들어오는 요청은 제외시킵니다.

```java
 @Override
    public void addInterceptors(final InterceptorRegistry registry) {
        registry.addInterceptor(authenticationInterceptor)
                .addPathPatterns("/carts/**")
                .excludePathPatterns("/signup"); // 회원가입 요청에 대해서는 authenticationInterceptor 을 타지 않도록 한다.
    }
```

이렇게 하면 인터셉터에서의 검증로직은 모두 로그인한 사용자에 대한 검증로직만 실행하면됩니다.
회원가입에 대한 검증 로직은 도메인의 값객체에서 예외가 발생할 것이기에 기대한 대로 동작하게 됩니다.

sonypark commented on src/main/java/cart/controller/CartApiController.java:
```diff
@@ -0,0 +1,50 @@
+        return cartService.findCartProducts(member);
+    }
+
+    @PostMapping("/carts/{productId}")
+    public ResponseEntity<Void> createCart(@AuthenticationPrincipal AuthInfo authInfo, @PathVariable final Long productId) {
+        final MemberEntity member = memberService.findByEmail(authInfo.getEmail());
+        final ProductEntity product = productService.findById(productId);
+        cartService.insert(member, product);
+        return ResponseEntity.ok().build();
+    }
```
컨트롤러에서 여러 서비스를 의존하여 호출하는건 문제 없습니다. 다만 현재 코드에서 문제가 되는 부분은 Entity가 프레젠테이션 레이어까지 노출되었다는 점입니다. Entity가 아닌 DTO로 포장하거나, cartService에서 memberService, productService를 의존해 사용하는게 더 나을 것 같습니다. 서비스가 서비스를 의존하는 부분에 대해서는 팀마다 룰이 다를 것 같습니다. 서비스간 의존은 허용하는 팀도 있을거고 서비스에서 서비스를 직접 의존하지 않고 DAO를 의존해서 사용하자고 컨벤션을 만든 팀이 있을 수 있습니다. (참고로 저희 팀은 후자 입니다.) 이 부분은 페어와 함께 상의하여 사용해보셔도 좋을 것 같네요 :)

Combi153 replied:
답변주셔서 감사합니다. 덕분에 의문이 해결되었습니다!
- Entity가 Controller 단에 노출된 것을 놓쳤네요. 이로 인해 계층을 나눠서 얻을 수 있는 장점 퇴색된 것 같습니다. Controller 에서 Entity가 사용되지 않도록 리팩터링하겠습니다.
- 서비스가 다른 서비스를 의존하도록 하는 것과, 서비스가 DAO를 의존하도록 하는 것 모두 장단점이 있을 것 같습니다. 이번에는 서비스가 DAO를 의존하도록 해보겠습니다. 서비스 간에 강결합이 유지보수 면에서 더 불리할 것으로 예상했기 때문입니다. 앞으로 두 방식 모두 사용해보면서 장단점을 몸소 체험해보고 저만의 기준을 가져보겠습니다!

sonypark commented on src/main/java/cart/controller/CartApiController.java:
```diff
@@ -0,0 +1,50 @@
+        this.cartService = cartService;
+        this.productService = productService;
+    }
+
+    @GetMapping("/carts")
+    public List<ResponseProductDto> readCart(@AuthenticationPrincipal AuthInfo authInfo) {
+        final MemberEntity member = memberService.findByEmail(authInfo.getEmail());
```
네 현재 아래 코드가 중복이네요. (이러면 인증이 필요한 모든 컨트롤러 코드에 같은 코드가 중복으로 들어가겠네요)
이 코드의 역할은 요청값에 있는 사용자 이메일을 통해 실제 존재하는 사용자인지를 판별하는 것인데요.
결국 인증의 역할입니다. Interceptor에 있는 인증 로직에서 하면 될 것 같습니다. (제가 interceptor를 사용할것을 말씀드린 의도이기도 했습니다.) 현재 Interceptor에서는 인증 로직은 없고 요청값에 대한 null 체크만 하고 있습니다. 이 부분을 Interceptor에 위임하면 좋을 것 같아요!

```java
final MemberEntity member = memberService.findByEmail(authInfo.getEmail());
```

Combi153 replied:
```java
@Component
public class AuthenticationInterceptor implements HandlerInterceptor {

    private final MemberService memberService;

    public AuthenticationInterceptor(final MemberService memberService) {
        this.memberService = memberService;
    }

    @Override
    public boolean preHandle(final HttpServletRequest request, final HttpServletResponse response, final Object handler) throws Exception {
        // header에서 인증 정보 추출하기
        final String header = request.getHeader(HttpHeaders.AUTHORIZATION);
        final AuthInfo authInfo = BasicAuthorizationExtractor.extract(header);
        // 인증 정보에 해당하는 회원이 있는지 확인하기
        if (memberService.hasMember(authInfo.getEmail(), authInfo.getPassword())) {
            return HandlerInterceptor.super.preHandle(request, response, handler);
        }
        throw new AuthorizationException("입력한 정보의 회원은 존재하지 않습니다.");
    }
}
```

Interceptor 는 preHandle 메서드에서 다음과 같은 로직을 수행합니다.
- header 에서 인증 정보 추출하기
- 인증 정보에 해당하는 회원이 있는지 확인하기(인증 정보에 맞는 회원이 없다면 예외)

>현재 Interceptor에서는 인증 로직은 없고 요청값에 대한 null 체크만 하고 있습니다.

소니께서 위와 같이 말씀해주셨는데, 저는 Interceptor에서 `인증 정보에 해당하는 회원이 있는지 확인하는 것`이 인증 로직이라고 생각했습니다. 한편, Controller에 있던 아래 메서드를 이미 인증이 끝난 상태에서 회원 객체를 생성하는 로직으로 생각했습니다.
```java
final MemberEntity member = memberService.findByEmail(authInfo.getEmail());
```

따라서 Interceptor에 해당 로직을 위임하지 않았습니다. 다만 리팩터링 과정에서 이 로직을 CartService 내부로 옮겼습니다. 회원 객체를 생성하는 이유는 비즈니스 로직을 수행하기 위함인데, 이는 CartService의 책임으로 판단했기 때문입니다. 리팩터링을 통해 Controller에 중복 코드를 없앨 수 있었습니다.

소니가 보시기에 여전히 Interceptor에 인증 로직이 부족하다고 느껴지시는지 궁금합니다!

sonypark replied:
인증로직은 충분합니다. 잘 작성해주셨어요! 다만 인증을 한 뒤 인증이 끝난 사용자에 대해서는 사용자 객체를 AuthenticationContext와 같은 컴포넌트에 저장해서 사용하면 불필요한 로직을 줄일 수 있지 않을까 싶어 제안드렸습니다.
현재는 인증이 완료된 사용자에 대해서 ArgumentResolver를 통해 AuthInfo 객체를 반환받아 다시 DB에서 Member를 조회하고 있습니다. 인터셉터에서도 Member를 조회해 인증하고, 서비스에서 다시 같은 정보로 Member를 조회해 사용하는 부분이 중복되고 불필요한 로직이라고 생각이 들었습니다. 인터셉터에서 DB 조회를 통해 인증을 통과하면 인증된 Member를  AuthenticationContext에 담아두고 ArgumentResolver에서 해당 Member를 get해서 반환하면 코드가 더 간결해지지 않을까 싶은데 후추의 생각은 어떠신가요?

sonypark reviewed (changes requested):
후추 안녕하세요!😃
마지막으로 개선하면 좋을 부분에 대해 코멘트 남겼으니 한 번 확인 부탁드려요!
(제 생각일뿐 동의하지 않으시면 꼭 반영하지 않으셔도 괜찮습니다. 후추의 의견도 남겨주세요~)

sonypark reviewed (changes requested):
후추 안녕하세요!😃
마지막으로 개선하면 좋을 부분에 대해 코멘트 남겼으니 한 번 확인 부탁드려요!
(제 생각일뿐 동의하지 않으시면 꼭 반영하지 않으셔도 괜찮습니다. 후추의 의견도 남겨주세요~)

Combi153 committed:
refactor: AuthInfo 에서 Email, Password 객체를 생성할 수 있도록 수정

Combi153 committed:
refactor: 회원 Dao에서 이메일 찾는 기능 수정

Combi153 committed:
refactor: 인증 시 예외 처리 수정

Combi153 committed:
feat: AuthInfo getter 추가

Combi153 committed:
feat: CartDao 예외 처리 추가

Combi153 committed:
refactor: MemberDao 예외 처리 로직 변경

Combi153 committed:
refactor: ProductDao 예외 처리 로직 변경

Combi153 committed:
refactor: CartService 로직 변경

Combi153 commented:
안녕하세요 소니!

후추입니다 😀 

뒤늦게 리뷰 요청을 드립니다!

리뷰해주신 부분에 대해 코드를 수정하고, 제 생각을 코멘트로 남겨두었습니다.

마지막까지 세심히 리뷰해주셔서 더 많이 성장할 수 있던 것 같습니다. 정말 감사드립니다 😃

sonypark reviewed (approved):
후추 안녕하세요. 리뷰어 소니입니다.
장바구니 미션 마무리하느라 고생하셨습니다.
매 PR마다 후추의 고민과 의견을 잘 정리해주셔서 참 좋았습니다.
테스트 코드도 꼼꼼히 작성하시고 도메인 로직도 깔끔하게 작성해주셨네요💯
다음 미션도 화이팅하세요!! 고생하셨어요!👏
